### ISSUE TITLE
twisted.web.client.HTTPClientFactory sends Host header (and others) more than once

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10943]
|<img alt="trickos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| trickos reported|
|-|-|
|Trac ID|trac#1382|
|Type|defect|
|Created|2005-12-27 22:12:53Z|
|Branch|https://github.com/twisted/twisted/tree/http-client-duplicate-headers-1382|

Attachments:

* [httpclientfactory.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c0f/c0f00c7e9572eb3f65756c0d2214d3e6192a8f83/b6d15ad16bf14de86caa8979e86d3dcedfdf4aac.patch) (903 bytes) - added by trickos on 2005-12-27 22:12:50Z - 
* [twisted-duplicated_host-test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c0f/c0f00c7e9572eb3f65756c0d2214d3e6192a8f83/06b3904a35194c7346547bd6e921d1a7572a1979.patch) (1859 bytes) - added by trickos on 2006-09-13 11:22:45Z - Diff including a test for duplicated HTTP host headers

<details><summary>Searchable metadata</summary>

```
trac-id__1382 1382
type__defect defect
reporter__trickos trickos
priority__high high
milestone__ 
branch__branches_http_client_duplicate_headers_1382 branches/http-client-duplicate-headers-1382
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1135721573000000 1135721573000000
changetime__1221258568000000 1221258568000000
version__ 
owner__ 
cc__jknight cc__trickos cc__oubiwann
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="trickos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|trickos commented|
|-|-|

```
#!html
<pre>
In HTTPPageGetter.connectionMade, some headers are sent before others, probably
to comply with broken proxies or help request dispatching. However, apart from
the 'content-length' one, all others may be sent once again if they belong to
the instance 'headers' attribute, which may happen with the 'Host' header for
instance. For headers not made of comma-separated values (like 'Host'), this is
forbidden by the RFC.

The attached patch lists the headers sent before all others in a 'first class'
dict, and avoid sending them again.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

This patch needs a unit test.

**Comment by twisted-trac** [comment]
|<img alt="trickos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|trickos commented|
|-|-|

Sorry for the delay,
I just added the test case for this issue. It can be executed with twisted 2.2 and 2.4, I could not find the time to test it against the dev branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [24737]) Branching to 'http-client-duplicate-headers-1382'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

Okay, fixed.  This whole area is kind of gross. :(  Maybe the new http client code will be super great and we can delete this soon.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @exarkun|
|-|-|

Fine enough, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [24817]) Merge http-client-duplicate-headers-1382

Author: exarkun
Reviewer: therve
Fixes: [#10943](https://github.com/twisted/twisted/issues/10943)

Fix `HTTPPageGetter` so that it does not incorrectly
duplicate certain headers when they are included in
the request header dictionary passed to the factory
for that protocol.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
